shell/platform/android/io/flutter/embedding/android/FlutterFragment.java

```diff
@@ -766,7 +766,15 @@ public String getCachedEngineId() {
    */
   @Override
   public boolean shouldDestroyEngineWithHost() {
-    return getArguments().getBoolean(ARG_DESTROY_ENGINE_WITH_FRAGMENT, false);
+    boolean explicitDestructionRequested = getArguments().getBoolean(ARG_DESTROY_ENGINE_WITH_FRAGMENT, false);
+    if (getCachedEngineId() != null) {
```

Comment:
This is not sufficient right? You'd probably want to check something like FlutterActivityAndFragmentDelegate.isFlutterEngineFromHost (perhaps package private) in case the engine came from provideFlutterEngine. You wouldn't want to kill that engine either.

Reply:
True. I added logic for provided engines and added tests. I had to expose some things for the tests due to our limited infrastructure, but hopefully it looks alright to you.